feat: add groups legend management

arcalys opened the pull request:
This is an attempt to integrate a groups legend into the "Who's online" block on the index page. It’s certainly not perfect and may need to be updated or improved, but for now, it’s working well for me.

### Added

- Added a new setting (defaults to 1) to control whether the groups legend is displayed.
- Added a new toggle in the groups management page to decide if a group should appear in the legend. Defaults to 1 for Administrators, Super Moderators, and Moderators.
- Added a new element in the Twig index view to display the groups legend above the members list, if enabled.

### Example

Index page:
![image](https://github.com/user-attachments/assets/f2dac55a-9222-442e-bad6-830612e2f8f1)

Setting in the ACP > Forum Home Options:
![image](https://github.com/user-attachments/assets/f8b23dc5-da98-4154-98e1-fe5143ae1e8f)

Group management:
![image](https://github.com/user-attachments/assets/0f91f44a-c3a8-4285-af1e-6a0da627fa7c)

Feedback and suggestions are welcome.

Commit:
feat: add groups legend management

dvz commented:
Would it be better to use the main `usergroups` datacache instead? This would eliminate the hassle of maintaining the derivative data.

We can use `format_name()`, or [Twig's `replace()`](https://twig.symfony.com/doc/3.x/filters/replace.html) (then without `|raw`) to deal with `{username}` dynamically.

The lang string was added after `forum_team` - we can similarly make the code more intuitive by adding the new lines for input/columns/defaults close to the equivalent _Show Team_ lines (adding everything before _team_ may even make sense, so that _General Options_ have increasing specificity); and the setting after `showwol`, since it's displayed under it (`<disporder>` will need correction).

arcalys commented:
> Would it be better to use the main `usergroups` datacache instead? This would eliminate the hassle of maintaining the derivative data.
> 
> We can use `format_name()`, or [Twig's `replace()`](https://twig.symfony.com/doc/3.x/filters/replace.html) (then without `|raw`) to deal with `{username}` dynamically.
> 
> The lang string was added after `forum_team` - we can similarly make the code more intuitive by adding the new lines for input/columns/defaults close to the equivalent _Show Team_ lines (adding everything before _team_ may even make sense, so that _General Options_ have increasing specificity); and the setting after `showwol`, since it's displayed under it (`<disporder>` will need correction).

Thanks a lot for the feedback!

I'm not sure why I didn't think of using the existing groups cache earlier — I've now updated everything to rely on that instead.

I also moved the `showinlegend` toggle to appear before `showforumteam`, and switched to using `format_name()` for rendering the group display. The general setting has been relocated next to `showwol`.

dvz commented on admin/modules/user/groups.php:
```diff
@@ -1070,6 +1071,10 @@
+	if($mybb->settings['showgroupslegend'] != 0)
+	{
+		$general_options[] = $form->generate_check_box("showinlegend", 1, $lang->show_in_legend, array("checked" => $mybb->input['showinlegend']));
+	}	
```
Would the missing input result in the option set to `0` for the group when the setting is disabled?

It may be fine to show the option regardless.

arcalys replied:
Correct, it would reset to 0 for a group that was previously shown in the legend, which would be awkward when reactivating the global setting. Removed the conditional and now display the option regardless of the setting's status.

Sama34 replied:
You may use a peeker

arcalys replied:
Is it necessary, or should the input be visible at all times? It doesn't bother me that it's always displayed, especially since I know disabling the global setting won't have any effect anyway.

Sama34 replied:
Always showing the input seems congruent to other behaviour, so no peeker is fine I think.

arcalys replied:
Ok then!

dvz commented:
Should the legend be displayed below the list of users instead?

arcalys commented:
Your choice, I don't have any preference.

dvz commented:
It seems customary, so let's move it below. Otherwise LGTM.

arcalys commented:
Updated:
![image](https://github.com/user-attachments/assets/e5877438-97c2-4768-9de9-952f91472059)

And rebased the PR to keep it clean.

arcalys commented:
Migration if PR is accepted:
```php
#pgsql
$db->add_column("usergroups", "showinlegend", "smallint NOT NULL default '0'");

# sqlite and mysql
$db->add_column("usergroups", "showinlegend", "tinyint(1) NOT NULL default '0'");
```

Sama34 commented on inc/themes/core.base/frontend/templates/index/index.twig:
```diff
@@ -32,6 +32,14 @@
                         {% if user.invisible == 1 and (mybb.usergroup.canviewwolinvis == 1 or user.uid == mybb.user.uid) %}*{% endif %}
                     {% endfor %}
                 </p>
+                {% if mybb.settings.showgroupslegend %}
```
```suggestion
                {% if mybb.settings.showgroupslegend and groups %}
```

Otherwise, some HTML will be added even if there are no groups to display.

Why do we need a global setting for this again ? I think the global setting to be unnecessary now.

arcalys replied:
If you want to disable the legend entirely and have many groups, it's probably easier to just toggle the global setting.

arcalys committed:
fix: show group legend only if groups are present

Co-authored-by: Omar G. <Sama34@users.noreply.github.com>
